src/consumers/metric-types/counter/counter.c

```diff
@@ -0,0 +1,121 @@
+#include <errno.h>
+#include <string.h>
+#include "../../../config-reader/config-reader.h"
+#include "../../../statsd-parsers/statsd-parsers.h"
+#include "../../../utils/utils.h"
+#include "../../consumers.h"
+#include "./counter.h"
+
+void init_counter_consumer(agent_config* config) {
+
+}
+
+/**
+ * Processes datagram struct into counter metric 
+ * @arg m - Metrics struct acting as metrics wrapper
+ * @arg datagram - Datagram to be processed
+ */
+void process_counter(metrics* m, statsd_datagram* datagram) {
+    counter_metric* counter = (struct counter_metric*) malloc(sizeof(counter_metric));
+    ALLOC_CHECK("Unable to allocate memory for placeholder counter record");
+    *counter = (struct counter_metric) { 0 };
+    if (find_counter_by_name(m, datagram->metric, &counter)) {
+        if (!update_counter_record(counter, datagram)) {
+            verbose_log("Throwing away datagram, semantically incorrect values.");
+            free_datagram(datagram);
+        }
+    } else {
+        if (check_metric_name_available(m, datagram->metric)) {
+            if (create_counter_record(datagram, &counter)) {
+                add_counter_record(m, counter);
+            } else {
+                verbose_log("Throwing away datagram, semantically incorrect values.");
+                free_datagram(datagram);
+            }
+        }
+    }
+}
+
+/**
+ * Writes information about recorded counters into file
+ * @arg out - OPENED file handle
+ * @return Total count of counters printed
+ */
+int print_counter_metric_collection(metrics* m, FILE* out) {
+    counter_metric_collection* counters = m->counters;
+    long int i;
+    for (i = 0; i < counters->length; i++) {
+        fprintf(out, "%s = %llu (counter)\n", counters->values[i]->name, counters->values[i]->value);
+    }
+    return counters->length;
+}
+
+/**
+ * Find counter by name
+ * @arg name - Metric name to search for
+ * @arg out - Placeholder counter metric
+ * @return 1 when any found
+ */
+int find_counter_by_name(metrics* m, char* name, counter_metric** out) {
+    counter_metric_collection* counters = m->counters;
+    long int i;
+    for (i = 0; i < counters->length; i++) {
+        if (strcmp(name, counters->values[i]->name) == 0) {
+            *out = counters->values[i];
+            return 1;
+        }
+    }
+    return 0;
+}
```

Comment:
For the record, this is O(n) but we already discussed that this needs to be stored in hash. Another option could be a (red-black) tree or other data structures, but honestly I'd go for a hash.

Reply:
Stored in hash by 085d60c

Reply:
Cleaned up some stuff and allowed same metric name to occupy multiple metric types. 17f257f